[FLINK-32549][network] Tiered storage memory manager supports ownership transfer for buffers

TanYuxin-tyx opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Currently, the accumulator is responsible for requesting all buffers, leading to an inaccurate number of requested buffers for each tier.
To address this issue, buffer ownership must be transferred from the accumulator to the tiers when writing them, which will enable the memory manager to maintain a correct number of requested buffers for different owners.


## Brief change log

  - *Tiered storage memory manager supports ownership transfer for buffers*
 
**Version 1 (Deprecated. Do not use this way to fix anymore)**:
1. In order to transfer ownership, added a new API `TieredStorageMemoryManager#transferBufferOwnership`.
2. If `TierProducerAgent#tryWrite` returns `true` (indicating the write is successful), then 
 the buffer has been written to the tier(once written, the buffer may be consumed or flushed immediately, so the buffer recycler can not be updated after `TierProducerAgent#tryWrite`), `TieredStorageProducerClientImpl#writeAccumulatedBuffer` can not update the buffer recycler anymore, so the updating buffer recycler process is moved into each tier.
3. Because the updating buffer recycler process is moved into each tier and the buffer recycler can not be updated after compression, the buffer compress process is also moved into each tier.
4. To use the right buffer recycler, added a new API `TieredStorageMemoryManager#getBufferRecycler`.

**Version 2**:
1. `Buffer` added a new API `setRecycler`
2. Added a new API for memory manager `TieredStorageMemoryManager#transferBufferOwnership`

## Verifying this change

This change added tests and can be verified as follows:

  - *Added tests to verify the buffer ownership transfer in `TieredStorageMemoryManagerImplTest`*
- *Added tests to verify getting the buffer recycler in `TieredStorageMemoryManagerImplTest`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

xintongsong commented:
@TanYuxin-tyx, I have 2 questions regarding the current approach.
1. The transfer of the buffer ownership seems to be divided into two steps: a) updating the buffer recycler, and b) updating the bookkeepings in memory manager. Currently, the two steps take place in different positions. Why is that necessary?
2. Why is updating of recycler not possible after the buffer being compressed?

TanYuxin-tyx commented:
@flinkbot run azure

TanYuxin-tyx commented:
@xintongsong Thanks for helping review. Based on the offline discussions, I have updated the change. In the new PR, the above issues no longer exist anymore. PTAL, thanks.

The change details are as follows.
1. Buffer added a new API `setRecycler`.
2. Added a new API for memory manager `TieredStorageMemoryManager#transferBufferOwnership`.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java:
```diff
@@ -276,6 +294,13 @@ private void recycleBuffer(Object owner, MemorySegment buffer) {
         decNumRequestedBuffer(owner);
     }
 
+    private BufferRecycler getOwnerBufferRecycler(Object owner) {
+        return memorySegment -> {
+            bufferPool.recycle(memorySegment);
+            decNumRequestedBuffer(owner);
+        };
+    }
```
This seems to be duplicated with `recycleBuffer`.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java:
```diff
@@ -41,6 +41,8 @@ class DiskCacheManager {
 
     private final int numSubpartitions;
 
+    private final TieredStorageMemoryManager memoryManager;
```
What is the purpose of this change? The filed seems never accessed after initialized.

TanYuxin-tyx replied:
Fixed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java:
```diff
@@ -200,6 +202,22 @@ public int numOwnerRequestedBuffer(Object owner) {
+    public void transferBufferOwnership(Object oldOwner, Object newOwner, Buffer buffer) {
+        if (!buffer.isBuffer()) {
+            return;
+        }
```
Moved the filter of `is null` outside the method.

reswqa commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java:
```diff
@@ -200,6 +201,20 @@ public int numOwnerRequestedBuffer(Object owner) {
+    public void transferBufferOwnership(Object oldOwner, Object newOwner, Buffer buffer) {
+        checkState(buffer.isBuffer(), "Only buffer supports transfer ownership.");
+        AtomicInteger numOldOwnerRequested = numOwnerRequestedBuffers.get(oldOwner);
+        if (numOldOwnerRequested == null) {
+            throw new RuntimeException("Failed to transfer buffer ownership for " + oldOwner);
+        }
```
This can be replaced by `checkNotNull`.

TanYuxin-tyx replied:
Use `decNumRequestedBuffer(oldOwner)` instead.

reswqa commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java:
```diff
@@ -87,6 +87,13 @@ public interface Buffer {
+     *
+     * @param bufferRecycler the new buffer recycler
+     */
+    void setRecycler(BufferRecycler bufferRecycler);
```
This is an unsafe operation, we may need add more descriptions of this method to avoid others accidentally introducing bugs. Especially for slice buffer.

TanYuxin-tyx replied:
Fixed. Added docs to describe the method.

reswqa reviewed (commented):
Thanks for the fix, this changes overall looks good to me. I only left some minor comments.

reswqa commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java:
```diff
@@ -196,8 +197,16 @@ public int getMaxNonReclaimableBuffers(Object owner) {
     public int numOwnerRequestedBuffer(Object owner) {
-        AtomicInteger numRequestedBuffer = numOwnerRequestedBuffers.get(owner);
-        return numRequestedBuffer == null ? 0 : numRequestedBuffer.get();
+        Integer numRequestedBuffer = numOwnerRequestedBuffers.get(owner);
+        return numRequestedBuffer == null ? 0 : numRequestedBuffer;
```
Can `numOwnerRequestedBuffers.getOrDefault(owner, 0)` simplify this?

reswqa reviewed (approved):
Thanks for the update, LGTM.
